pkg/controller/sparkapplication/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,162 @@
+	volcanoClient   volcanoclient.Interface
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string) bool {
+		return scheduler != nil && *scheduler == v.Name()
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName) && checkScheduler(app.Spec.Driver.SchedulerName)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) OnSubmitSparkApplication(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	if app.Spec.Executor.Annotations == nil {
+		app.Spec.Executor.Annotations = make(map[string]string)
+	}
+
+	if app.Spec.Driver.Annotations == nil {
+		app.Spec.Driver.Annotations = make(map[string]string)
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupForClientAPP(app)
+	} else if app.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupForClusterAPP(app)
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClientAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	if _, ok := app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, *app.Spec.Executor.Instances); err == nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClusterAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We need both mark Driver and Executor when submitting
+	//NOTE: Although we only have one pod for Spark Driver, we still manage it into PodGroup,since it can
+	//utilize other advanced features in volcano, for instance, namespace resource fairness.
+	if _, ok := app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, 1); err != nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+			app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) OnSparkDriverPodScheduled(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	// Update PodGroup MinMember if required
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return app, v.syncPodGroup(app, 1+*(app.Spec.Executor.Instances))
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) getAppPodGroupName(app *v1beta1.SparkApplication) string {
+	return fmt.Sprintf("sparkapplication-%s-podgroup", app.Name)
```

Comment:
Let's remove the prefix `sparkapplication` to reduce the chance of the name being too long.